packages/provider/wrapper.js

```diff
@@ -32,6 +32,43 @@ module.exports = {
       const originalSendAsync = provider.sendAsync.bind(provider);
       provider.sendAsync = this.send(originalSendAsync, preHook, postHook);
     }
+
+    // EIP-1193 provider wrapping
+    if (
+      typeof provider.request === "undefined" &&
+      typeof provider.send !== "undefined"
+    ) {
+      provider.request = async function ({ method, params }) {
+        let id = 0;
+        return await new Promise((accept, reject) => {
+          provider.send(
+            {
+              jsonrpc: "2.0",
+              id: ++id,
```

Comment:
`id` should probably be declared outside the scope of `provider.request` because it will always be 1, at this point.

Reply:
probably better just to use the timestamp, because otherwise there's risk that someone will intersperse `.request` with `.send` and the IDs might collide.

Reply:
It was changed from `timestamp` to `id` because timestamps are a) non-deterministic, and b) not granular enough -- this matters if we ever want to switch to websockets, which matches responses to requests by `id`.

Reply:
Ah thank you for the context @davidmurdoch! So we do have a race condition and need to track the IDs across `.send` and `.request` invocations?

Reply:
If we switch to WebSockets _and_ send parallel requests, potentially yes. This is what the JSON-RPC `id` field is for. We could also use a UUID, but those are slower to generate and aren't deterministic (not that determinism is actually important here -- it's just a nice to have). In early development of forking websocket URL support I ran into this race condition.

HTTP doesn't care though, and that's all that Truffle can use internally today, last I checked, so this is only for a future where Truffle switches to using a WebSocket provider.